src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,157 @@
+        if (initialCapacity > 0) {
+            this.elementData = new Object[initialCapacity];
+        } else if (initialCapacity == 0) {
+            this.elementData = EMPTY_ELEMENTDATA;
+        } else {
+            throw new IllegalArgumentException("Illegal Capacity: "
+                    + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < this.elementData.length) {
+            elementData[size] = value;
+            size++;
+        } else {
+            add(value, size);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        rangeCheckForAdd(index);
+        if (size == this.elementData.length) {
+            elementData = growIfArrayFull();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                    size - index);
+        elementData[index] = value;
+        size++;
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + ", Size: " + size;
+    }
+
+    private Object[] growIfArrayFull(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity > 0 || elementData != DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            int newCapacity = (int)(oldCapacity * CAPACITY_GROWTH_RATIO);
+            Object[] elementDataNew = new Object[newCapacity];
+            System.arraycopy(elementData,0, elementDataNew, 0, elementData.length);
+            return elementDataNew;
+        }
+        return new Object[Math.max(DEFAULT_CAPACITY, minCapacity)];
+    }
 
+    private Object[] growIfArrayFull() {
+        return growIfArrayFull(size + 1);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] elementDataNew = new Object[list.size()];
+        for (int i = 0; i < elementDataNew.length; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexInBounds(index)) {
+            return (T) elementData[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private boolean isIndexInBounds(int index) {
+        return index <= size - 1 && index >= 0;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isIndexInBounds(index)) {
+            elementData[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
```

Comment:
The index check logic is repeated in 'get', 'set', and 'remove' methods. Consider refactoring this logic into a single method to avoid code duplication and adhere to the DRY principle.